### ISSUE TITLE
forall inside record doesn't compile due to const ref `this`

### ISSUE THREAD

**Issue by BryantLam** [chapel-lang/chapel#13297]
Probably a duplicate of #12371
Bug. Design, compiler.

This record initializes an array field with `forall`. The three loops don't work for different reasons. Loops 2 and 3 should be legal. Loop 1 is debatable since record type `this` has `const ref` intent; I'm not sure what the consequences are of allowing that pattern since it [appears safe inside the record](https://github.com/chapel-lang/chapel/issues/12371#issuecomment-468853793).

```chapel
record R {
    var data: [1..8] int;
    
    // Any member function. Using `postinit` for code brevity.
    proc postinit() {
        forall i in data.domain { // `this` is const ref
            data[i] = i;
        }
        
        forall i in data.domain with (ref this) {
            data[i] = i; // doesn't recognize as `this.data`
        }
        
        forall i in data.domain with (ref this.data) { // syntax problem
            this.data[i] = i;
        }
    }
}

proc main() {
    var r = new R();
}
```
`chpl version 1.19.0`

**Edit:** Whoops! Fixed the loop body so it's not a blatant data race. Thanks @cassella!

**Comment by cassella** [comment]
Loop 2.5 works,

```chapel
        forall i in data.domain with (ref this) {
            this.data = i;
        }
```

(Also works with `this.data[i] = i`.)

**Comment by bradcray** [comment]
Tagging @vasslitvinov on this issue.

I believe the third case is not ntended to be supported, at least currently:

>         forall i in data.domain with (ref this.data) { // syntax problem

Specifically, a loop's with-clause is only intended to express how outer-scope identifiers should be passed through to the tasks, not arbitrary expressions.

The first two cases look to me like they should work, though I'd expect (and prefer) that someone just to be able to write the first case.  My intuition is that for `forall` loops (and other parallel constructs) in methods like this, the intents should essentially be able to ignore `this` and just operate on the field types themselves.  That is, we don't think about the intent for `this` when considering `data` within the loop, we just think about the intent for `data` (as a record, its passed by reference).  Vass, what do you think?

**Comment by vasslitvinov** [comment]
Yeah, these are bugs, except for `with (ref this.data)`, which we discussed earlier.

An unrelated bug: postinit() cannot modify `this` when invoked for default-initialization of a `const`, ex. `const r: R;`.

**Comment by bradcray** [comment]
Is it OK if I assign this to you, Vass?

**Comment by vasslitvinov** [comment]
Thinking some more about this proposal:

> My intuition is that for forall loops (and other parallel constructs) in methods like this, the intents should essentially be able to ignore this and just operate on the field types themselves. 

If there is an explicit intent on `this`, should the field access abide by that intent or go by the field type? Should field accesses behave differently whether the field name is qualified with `this` or not, as in `data` vs. `this.data` below? Does the answer depend on whether or not there is an explicit intent on `this` in the with-clause? For example:

```chpl
record R {
  var data: [1..8] int;
  proc anyMemberFunction() {
    forall i in data.domain with (const in this) {
        data[i] = i;  // is `data` treated as part of the shadow variable for `this`
                      // or as the outer this's `this.data` passed by the default intent?
        this.data[i] = i;  // what if there is an explicit `this.` ?
    }
    forall i in data.domain {
        data[i] = i;       // current proposal: pass by ref since this is an array
        this.data[i] = i;  // how to think about this case?
    }
  }       
}
```

@bradcray @BryantLam what do you think?

**Comment by bradcray** [comment]
Offline, I suggested that `this` not be valid as a task intent (at least currently) as a means of dodging this question and avoiding potential confusion.  Symmetrically, variables named `this` can't be created since it's a reserved word.

For the case of `this.data` vs. `data` I suggested that they be treated identically.

**Comment by vasslitvinov** [comment]
#13463 will disallow 'this' in explicit task/forall intents.

Inside a forall loop body, accesses to 'this', other than field accesses, will still be implicit accesses to 'this' shadow variable, which will be passed by the default intent, i.e. 'const ref'. So, for example, an integer field of a record will be replaced with a 'const in' shadow variable if accessed directly in the loop body, while still reference the field in 'this' when accessed in a method invoked on 'this'.

For now, we are not providing a mechanism to override the default intent for 'this' or its fields. We will consider doing so if/when there is a need.

**Comment by vasslitvinov** [comment]
To illustrate the semantics as implemented in #13479, consider this template:
```chpl
record R {
  (var | const) myField: fieldType;

  proc myProc() {
    forall .... {
       ... (read | write) myField ...;
    }
}

(var | const) r: R;
r.myProc();
```

W.r.t. the forall loop in `myProc`, the field `this.myField` is be considered a variable passed into the loop by the default intent. Here is what happens depending on the var/const and read/write choices and on the default forall intent for `fieldType`.

This case is allowed with #13479 whereas it was not before:

* **The forall loop performs a write, `myField` and `r` are "var", the default intent for `fieldType` is "ref"** i.e. it is an array type. --> This is now allowed. `myProc` takes its receiver by `ref` intent.

The following behaviors are unchanged from before #13479:

* **The forall loop performs a write, other choices are not covered by the above case.** --> These cases are illegal and cause compiler errors. For example, when `myField` or `r` are declared "const" or `fieldType` is int or a domain.

* **The forall loop performs a read.** --> myProc takes its receiver by `const ref` and all combinations of var/const on `r` and `myField` are legal.

The following behavior is changed with #13479:

* When the default forall intent for `fieldType` is `const in`, for example for numeric types, references to `myField` within the loop are implicitly to a local copy. This eliminates remote references in some cases. If `this.myField` changes concurrently with the forall loop, the change is not observable within the loop.

  Cf. Prior to #13479, accesses to`myField` within the forall loop used to be essentially references to `this.myField` directly, with the possibility of being remote-value-forwarded by compiler optimizations.

**Comment by vasslitvinov** [comment]
For now, resolve this via #13479.
